### PR TITLE
Implement shipping manifest

### PR DESCRIPTION
**Pull request by uzayr**

### CODE REVIEW

File: __init__.py [hunk 1]
```diff
@@ -19,10 +19,11 @@
     ApplyShippingSelectRate, ApplyShipping
 from configuration import PartyConfiguration
 from log import CarrierLog
-
+from manifest import Manifest
 
 def register():
     Pool.register(
+        Manifest,
```
Comment (riteshshrv) [inline __init__.py]: make this the last entry

File: manifest.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
+"""
+    Shipping manifest is the end of the day document
+    which is usually given to driver or carrier provider.
+    It lists all the shipments ready to ship.
+    manifest.py
```
Comment (riteshshrv) [inline manifest.py]: remove all of this doc string
only `manifest.py` with license details

File: manifest.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    """
+        Manifest Model for shipping.
+    """
+    __name__ = "shipping.manifest"
+
+    carrier = fields.Many2One("carrier", "Carrier")
+
+    warehouse = fields.Many2One(
+        "stock.location",
+        "Warehouse Location",
```
Comment (riteshshrv) [inline manifest.py]: only `Warehouse`

File: manifest.xml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            <field name="name">shipping_manifest_form</field>
+        </record>
+        <record model="ir.ui.view" id="ship_manifest_view_tree">
+            <field name="model">shipping.manifest</field>
+            <field name="type">tree</field>
+            <field name="name">shipping_manifest_tree</field>
+        </record>
+
+        <record model="ir.action.act_window" id="act_shipping_manifest_form">
+            <field name="name">Manifesto</field>
```
Comment (riteshshrv) [inline manifest.xml]: typo?

File: manifest.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+"""
+
+from trytond.model import fields, ModelView, ModelSQL
+from trytond.pool import PoolMeta
+
+__metaclass__ = PoolMeta
+__all__  = ["Manifest"]
+
+class Manifest(ModelSQL, ModelView):
+    """
```
Comment (prakashpp) [inline manifest.py]: in single line

File: manifest.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        Manifest Model for shipping.
+    """
+    __name__ = "shipping.manifest"
+
+    carrier = fields.Many2One("carrier", "Carrier")
```
Comment (prakashpp) [inline manifest.py]: required=true, select=true

File: manifest.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    warehouse = fields.Many2One(
+        "stock.location",
```
Comment (prakashpp) [inline manifest.py]: required=True, select=True

File: manifest.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        "Warehouse Location",
+        domain=[("type", "=", "warehouse")]
+    )
+
+    shipments = fields.One2Many(
+        "stock.shipment.out",
+        "carrier",
+        "Shipments",
+        domain=[("carrier", "=", "shipment.carrier")]
```
Comment (prakashpp) [inline manifest.py]: domain=[('carrier', '=', Eval('carrier')]
depends=['carrier']

File: manifest.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    )
+
+    state = fields.Selection([
+        ("open", "Open"),
+        ("closed", "Closed")
+        ], "State", required=True, select=True
+    )
```
Comment (prakashpp) [inline manifest.py]: where are buttons to open and close?

File: manifest.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+__metaclass__ = PoolMeta
+__all__ = ["ShippingManifest"]
+STATES = {
+    "readonly": Eval("state") == "closed"
+}
+
+
+class ShippingManifest(Workflow, ModelSQL, ModelView):
+    """Manifest Model for shipping."""
```
Comment (riteshshrv) [inline manifest.py]: """
......
"""

Comment (uzayr) [inline manifest.py]: screen shot 2016-04-22 at 16 41 01

File: manifest.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    )
+
+    state = fields.Selection([
+        ("open", "Open"),
+        ("closed", "Closed")
+        ], "State", required=True, select=True, readonly=True
+    )
+
+    object_id = fields.Char("Object ID", readonly=True)
+    status = fields.Selection([
```
Comment (riteshshrv) [inline manifest.py]: remove this field; its not required here

File: manifest.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        depends=["carrier"]
+    )
+
+    state = fields.Selection([
+        ("open", "Open"),
+        ("closed", "Closed")
+        ], "State", required=True, select=True, readonly=True
+    )
+
+    object_id = fields.Char("Object ID", readonly=True)
```
Comment (riteshshrv) [inline manifest.py]: i don't think its required

**Comment by riteshshrv** [comment]
screenshot?

**Comment by uzayr** [comment]
screen shot 2016-04-22 at 18 54 47
screen shot 2016-04-22 at 18 55 07
https://3.basecamp.com/3194017/buckets/421793/todos/108878572

**Comment by riteshshrv** [comment]
LGTM

File: manifest.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+"""
+
+from datetime import datetime
+from trytond.model import fields, ModelView, ModelSQL, Workflow
+from trytond.pool import PoolMeta
+from trytond.pyson import Eval
+
+__metaclass__ = PoolMeta
+__all__ = ["ShippingManifest"]
+STATES = {
```
Comment (riteshshrv) [inline manifest.py]: where is this being used??

**Comment by uzayr** [comment]
screen shot 2016-04-26 at 19 15 49

File: manifest.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+__metaclass__ = PoolMeta
+__all__ = ["ShippingManifest"]
+
+
+class ShippingManifest(Workflow, ModelSQL, ModelView):
+    "Manifest Model for shipping."
+    __name__ = "shipping.manifest"
+
+    _rec_name = 'object_id'
```
Comment (prakashpp) [inline manifest.py]: what is this?

Comment (prakashpp) [inline manifest.py]: Remove object id

File: manifest.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+            "readonly": Eval("state") == "closed"
+        },
+        domain=[
+            ('carrier', '=', Eval('carrier')),
+            ('state', '=', 'done'),
+            ('tracking_number', '!=', None)
+        ],
+        add_remove=[
+            ('carrier', '=', Eval('carrier')),
+            ('shippo_manifest', '=', None),
```
Comment (prakashpp) [inline manifest.py]: manifest. Add a field to shipment

File: manifest.py (hunk 12 above)
Comment (prakashpp) [inline manifest.py]: remove this field

File: manifest.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+    state = fields.Selection([
+        ("open", "Open"),
+        ("closed", "Closed")
+        ], "State", required=True, select=True, readonly=True
+    )
+
+    object_id = fields.Char("Object ID", readonly=True)
+
+    open_date = fields.DateTime("Open Date", readonly=True, required=True)
```
Comment (prakashpp) [inline manifest.py]: remove required.

File: manifest.py [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+
+    @classmethod
+    def __setup__(cls):
+        super(ShippingManifest, cls).__setup__()
+
+        cls._transitions |= set((
+            ('open', 'closed'),
+        ))
+
+        cls._error_messages.update({
```
Comment (prakashpp) [inline manifest.py]: remove

**Comment by prakashpp** [comment]
Opened https://github.com/fulfilio/trytond-shipping/pull/92

**Pull request closed without merging**